metadata-ingestion/src/datahub/ingestion/transformer/dataset_domain.py

```diff
@@ -0,0 +1,137 @@
+from typing import Callable, List, Optional, Union
+
+from datahub.configuration.common import (
+    ConfigurationError,
+    KeyValuePattern,
+    Semantics,
+    SemanticsTransformerConfigModel,
+)
+from datahub.configuration.import_resolver import pydantic_resolve_key
+from datahub.emitter.mce_builder import Aspect
+from datahub.ingestion.api.common import PipelineContext
+from datahub.ingestion.graph.client import DataHubGraph
+from datahub.ingestion.transformer.dataset_transformer import DatasetDomainTransformer
+from datahub.metadata.schema_classes import DomainsClass
+
+
+class AddDatasetDomainConfig(SemanticsTransformerConfigModel):
+    get_domains_to_add: Union[
+        Callable[[str], DomainsClass],
+        Callable[[str], DomainsClass],
+    ]
+
+    _resolve_domain_fn = pydantic_resolve_key("get_domains_to_add")
+
+
+class SimpleDatasetDomainConfig(SemanticsTransformerConfigModel):
+    domain_urns: List[str]
+
+
+class PatternDatasetDomainConfig(SemanticsTransformerConfigModel):
+    domain_pattern: KeyValuePattern = KeyValuePattern.all()
+
+
+class AddDatasetDomain(DatasetDomainTransformer):
+    """Transformer that adds domains to datasets according to a callback function."""
+
+    ctx: PipelineContext
+    config: AddDatasetDomainConfig
+
+    def __init__(self, config: AddDatasetDomainConfig, ctx: PipelineContext):
+        super().__init__()
+        self.ctx = ctx
+        self.config = config
+        if self.config.semantics == Semantics.PATCH and self.ctx.graph is None:
```

Comment:
Also resolve domains to fully qualified domain urns. 
See https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py#L513 for how this is done in other places.